specs/light_client/merkle_proofs.md

```diff
@@ -116,72 +199,69 @@ x x . . . . x *
-            index_queue.append(index // 2)
-        i += 1
-    return (indices == []) or (1 in tree and tree[1] == root)
+```
+def get_branch_indices(tree_index: GeneralizedIndex) -> List[GeneralizedIndex]:
+    """
+    Get the generalized indices of the sister chunks along the path from the chunk with the
+    given tree index to the root.
+    """
+    o = [generalized_index_sibling(tree_index)]
+    while o[-1] > 1:
+        o.append(generalized_index_sibling(generalized_index_parent(o[-1])))
+    return o[:-1]
+
+def get_helper_indices(indices: List[GeneralizedIndex]) -> List[GeneralizedIndex]:
+    """
+    Get the generalized indices of all "extra" chunks in the tree needed to prove the chunks with the given
+    generalized indices. Note that the decreasing order is chosen deliberately to ensure equivalence to the
+    order of hashes in a regular single-item Merkle proof in the single-item case.
+    """
+    all_indices = set()
+    for index in indices:
+        all_indices = all_indices.union(set(get_branch_indices(index) + [index]))
+    
+    return sorted([
+        x for x in all_indices if not
+        (generalized_index_child(x, 0) in all_indices and generalized_index_child(x, 1) in all_indices) and not
+        (x in indices)
+    ])[::-1]
 ```
 
-## MerklePartial
-
-We define:
-
-### `SSZMerklePartial`
-
+Now we provide the Merkle proof verification functions. First, for single item proofs:
 
 ```python
-{
-    "root": "bytes32",
-    "indices": ["uint64"],
-    "values": ["bytes32"],
-    "proof": ["bytes32"]
-}
+def verify_merkle_proof(leaf: Hash, proof: Sequence[Hash], index: GeneralizedIndex, root: Hash) -> bool:
+    assert len(proof) == get_generalized_index_length(index)
+    for i, h in enumerate(proof):
+        if get_generalized_index_bit(index, i):
+            leaf = hash(h + leaf)
+        else:
+            leaf = hash(leaf + h)
+    return leaf == root
 ```
 
-### Proofs for execution
+Now for multi-item proofs:
 
-We define `MerklePartial(f, arg1, arg2..., focus=0)` as being a `SSZMerklePartial` object wrapping a Merkle multiproof of the set of nodes in the hash tree of the SSZ object `arg[focus]` that is needed to authenticate the parts of the object needed to compute `f(arg1, arg2...)`.
+```python
+def verify_merkle_multiproof(leaves: Sequence[Hash], proof: Sequence[Hash], indices: Sequence[GeneralizedIndex], root: Hash) -> bool:
+    assert len(leaves) == len(indices)
+    helper_indices = get_helper_indices(indices)
+    assert len(proof) == len(helper_indices)
+    objects = {
+        **{index:node for index, node in zip(indices, leaves)},
+        **{index:node for index, node in zip(helper_indices, proof)}
+    }
+    keys = sorted(objects.keys())[::-1]
+    pos = 0
+    while pos < len(keys):
+        k = keys[pos]
+        if k in objects and k ^ 1 in objects and k // 2 not in objects:
+            objects[k // 2] = hash(objects[k & -2] + objects[k | 1])
```

Comment:
```suggestion
            objects[k // 2] = hash(objects[(k | 1) ^ 1] + objects[k | 1])
```
Not sure about this, python doesn't seem to have a `&^` operator, and negative integers for bit-masks feel a bit much for a spec (although I enjoy bit-magic). This makes it clear that they are sibling indices at least.